src/modules/storage/LedgerStorage.ts

```diff
@@ -119,4 +130,107 @@ export class LedgerStorage extends Storages
         });
     }
 
+    /**
+     * Put a enrollment to database
+     * @param data a enrollment data
+     * @param callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     * The first argument is an error object.
+     */
+    public putEnrollment (data: any, callback: (err: Error | null) => void)
+    {
+        if (
+                (data == null) ||
+                (data.block_height == undefined) ||
+                (data.enrollment_index == undefined) ||
+                (data.utxo_key == undefined) ||
+                (data.random_seed == undefined) ||
+                (data.cycle_length == undefined) ||
+                (data.enroll_sig == undefined)
+        ) {
+            if (callback != null)
+                callback(new Error("Parameter validation failed."));
+            return;
+        }
+
+        var sql =
+        `INSERT INTO enrollments
+            (block_height, enrollment_index, utxo_key, random_seed, cycle_length, enroll_sig)
+        VALUES
+            (?, ?, ?, ?, ?, ?)`;
+        this.db.run(sql,
+            [
+                data.block_height,
+                data.enrollment_index,
+                data.utxo_key,
+                data.random_seed,
+                data.cycle_length,
+                data.enroll_sig
+            ], (err: Error | null) =>
+        {
+            if (callback != null)
+                callback(err);
+        });
+    }
+
+    /**
+     * Puts all enrollments
+     * @param data: block header JSON object
+     * @param Callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     */
+    public putAllEnrollments (header: any, callback: (err: Error | null) => void)
+    {
+        var idx: number = 0;
+        var doPut = () =>
+        {
+            if (idx >= header.enrollments.length)
+            {
+                return;
+            }
+
+            let enrollment: any = header.enrollments[idx];
+            enrollment.block_height = header.height.value;
+            enrollment.enrollment_index = idx;
+
+            this.putEnrollment(enrollment, (err: Error | null) =>
+            {
+                if (err != null)
+                {
+                    idx++;
+                    doPut();
+                }
+                else
+                {
+                    if (callback != null)
+                        callback(err);
+                    else
+                        return;
+                }
+            });
+        }
+        doPut();
+    }
```

Comment:
Why does a foreach have to be 30 lines ?

Reply:
Ah you are just moving the code

Reply:
More importantly, why is this recursive?

Reply:
You're watching a feast of Asynchronous Bracelet.
If I find and write a good asynchronous library, I think the code will be simple.
I don't have enough time to apply it.

Reply:
I think this is the wrong place to write asynchronous code.

Reply:
@AndrejMitrovic 
I didn't understand what you meant.